### PR TITLE
add gradient checkpointing 

### PR DESCRIPTION
**Pull request by SeunghyunSEO**
- add gradient checkpointing
- set_seed
- remove per_device hard coding when head_width > 32
- update readme (deprecated arguments)

(please ignore wrong branch name, my bad) 

Gradient checkpointing (grad ckpt) is tested on 4x 80GB A100.  
I set seed as 42 for both exps w/ and w/o gradient checkpointing and snapshot the 4th batch of each exps.
here is my bash script and profiling results (inpus shape is printed as well).

```
# w/ grad ckpt
deepspeed --num_gpus 4 run_trainer.py --learning_rate 1e-4 --head_width 128 --run_name "test" --per_device_train_batch_size 128 --print_profile_results

# w/o grad ckpt
deepspeed --num_gpus 4 run_trainer.py --learning_rate 1e-4 --head_width 128 --run_name "test" --per_device_train_batch_size 128 --print_profile_results --gradient_checkpointing False
```

You can check the loss of both exps are same and exp w/ grad ckpt consumes 59GB as max reserved, but 70GB for exp w/o grad.
However, idk why training w/o grad ckpt is slower than another one.
(maybe because of the memory pressure? because when batch size is 128 and num_gpus=4, optimizer step happens in every 4 iteration)

- w/ grad ckpt

```python
[2024-03-08 16:55:14,292] [INFO] [memory_profile_utils.py:35:get_nvidia_gpu_memory]                                                                                                                                                                         
    nvidia-smi info                                                                                                                                                                                                                                         
    [62053, 62341, 62341, 62053]                                                                                                                                                                                                                            
[2024-03-08 16:55:14,413] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 3 / device: cuda:3                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.06 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 59096.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 51479.94MB / 59098.00MB                                                                                                                                                                                                   
[2024-03-08 16:55:14,446] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 1 / device: cuda:1                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.06 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 59096.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 51479.94MB / 59098.00MB                                                                                                                                                                                                   
[2024-03-08 16:55:14,458] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 2 / device: cuda:2                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.06 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 59096.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 51479.94MB / 59098.00MB                                                                                                                                                                                                   
[2024-03-08 16:55:14,473] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 0 / device: cuda:0                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.06 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 59096.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 51479.94MB / 59098.00MB                                                                                                                                                                                                   
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
[2024-03-08 16:55:15,920] [INFO] [run_trainer.py:68:train] loss : 10.528924942016602                                                                                                                                                                        
[2024-03-08 16:55:32,394] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | optimizer_step: 13770.43                                                                                                                                                     
[2024-03-08 16:55:32,396] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd_microstep: 990.59 | bwd_microstep: 2566.82 | bwd_inner_microstep: 349.25 | bwd_allreduce_microstep: 2217.48 | step_microstep: 13908.51                                   
[2024-03-08 16:55:32,396] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd: 3732.68 | bwd: 5200.69 | bwd_inner: 1943.69 | bwd_allreduce: 3256.66 | step: 13908.68 
```

- w/o grad ckpt

```python
[2024-03-08 16:52:36,276] [INFO] [memory_profile_utils.py:35:get_nvidia_gpu_memory]                                                                                                                                                                         
    nvidia-smi info                                                                                                                                                                                                                                         
    [73335, 73623, 73623, 73335]                                                                                                                                                                                                                            
[2024-03-08 16:52:36,414] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 1 / device: cuda:1                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.04 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 70378.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 63791.94MB / 70380.00MB                                                                                                                                                                                                   
[2024-03-08 16:52:36,437] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 2 / device: cuda:2                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.04 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 70378.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 63791.94MB / 70380.00MB                                                                                                                                                                                                   
[2024-03-08 16:52:36,444] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 3 / device: cuda:3                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.04 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 70378.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 63791.94MB / 70380.00MB                                                                                                                                                                                                   
[2024-03-08 16:52:36,550] [INFO] [memory_profile_utils.py:64:gpu_memory_plot_helper]                                                                                                                                                                        
    rank: 0 / device: cuda:0                                                                                                                                                                                                                                
    CPU Virtual Memory:  used = 149.04 GB, percent = 7.4%                                                                                                                                                                                                   
    Allocated / Reserved: 13585.61MB / 70378.00MB                                                                                                                                                                                                           
    Max Allocated / Max Reserved: 63791.94MB / 70380.00MB                                                                                                                                                                                                   
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
x.size() torch.Size([128, 512, 512])                                                                                                                                                                                                                        
[2024-03-08 16:52:38,294] [INFO] [run_trainer.py:68:train] loss : 10.528924942016602                                                                                                                                                                        
[2024-03-08 16:52:55,712] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | optimizer_step: 14793.44                                                                                                                                                     
[2024-03-08 16:52:55,713] [WARNING] [stage3.py:2069:step] 1 pytorch allocator cache flushes since last step. this happens when there is high memory pressure and is detrimental to performance. if this is happening frequently consider adjusting settings 
to reduce memory consumption. If you are unable to make the cache flushes go away consider adding get_accelerator().empty_cache() calls in your training loop to ensure that all ranks flush their caches at the same time                                  
[2024-03-08 16:52:55,715] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd_microstep: 1139.51 | bwd_microstep: 2530.78 | bwd_inner_microstep: 260.12 | bwd_allreduce_microstep: 2270.55 | step_microstep: 14889.10                                  
[2024-03-08 16:52:55,716] [INFO] [logging.py:96:log_dist] [Rank 0] time (ms) | fwd: 3447.81 | bwd: 5348.57 | bwd_inner: 2118.37 | bwd_allreduce: 3229.82 | step: 14889.30 
```

### CODE REVIEW

**Comment by SeunghyunSEO** [comment]
now plz do large sacle thing, not small scale thing @cloneofsimo XD
(show us exp results faster and faster)

**Comment by cloneofsimo** [comment]
LGTM bro
